### PR TITLE
Merge headers with default of RequestOptions DI

### LINKED ISSUE #309: JWT should take default RequestOptions in account when setting headers
I provide some headers via defaultOpts:
```
export function requestOptionsFactory(): RequestOptions {
  const requestOptions: RequestOptions = new BaseRequestOptions();
  requestOptions.headers.set('Content-Type', 'application/json');
  return requestOptions;
}
```

My provided RequestOptions.headers is null, the method setGlobalHeaders creates a new instance of Header and overwrite so completly my provided default Headers.
```
  private mergeOptions(providedOpts: RequestOptionsArgs, defaultOpts?: RequestOptions) {
    let newOptions = defaultOpts || new RequestOptions();
    if (this.config.globalHeaders) {
      this.setGlobalHeaders(this.config.globalHeaders, providedOpts);
    }

    newOptions = newOptions.merge(new RequestOptions(providedOpts));

    return newOptions;
  }
```

### PR DESCRIPTION
**Pull request by CSchulz**
fixes #309

### CODE REVIEW

**Comment by escardin** [comment]
Can you add some tests proving this fixes the issue?

File: angular2-jwt.ts [hunk 1]
```diff
@@ -109,7 +109,19 @@ export class AuthHttp {
       this.setGlobalHeaders(this.config.globalHeaders, providedOpts);
     }
 
-    newOptions = newOptions.merge(new RequestOptions(providedOpts));
+    let providedOptions = new RequestOptions(providedOpts);
+    if (newOptions.headers && newOptions.headers.keys().length > 0) {
```
Comment (escardin) [inline angular2-jwt.ts]: checking keys().length is not needed. The forEach won't do anything if there are no keys

File: angular2-jwt.ts [hunk 2]
```diff
@@ -109,7 +109,19 @@ export class AuthHttp {
...
+
+        // set only not existing header keys
+        if (!providedOptions.headers.has(name)) {
+          let headerValue: string = (header as any)[Object.keys(header)[0]];
+          providedOptions.headers.set(name, headerValue);
+        }
+      });
+    }
+
+    newOptions = newOptions.merge(providedOptions);
```
Comment (escardin) [inline angular2-jwt.ts]: This whole block seems really awkward.

for every header in newOptions, set that header in providedOptions if it isn't already set, then merge provided options into new options.

Which headers is kept?

File: angular2-jwt.ts [hunk 3]
```diff
@@ -109,7 +109,19 @@ export class AuthHttp {
...
     }
 
-    newOptions = newOptions.merge(new RequestOptions(providedOpts));
+    let providedOptions = new RequestOptions(providedOpts);
+    if (newOptions.headers && newOptions.headers.keys().length > 0) {
+      newOptions.headers.forEach((header: Object, name: string) => {
+
+        // set only not existing header keys
+        if (!providedOptions.headers.has(name)) {
+          let headerValue: string = (header as any)[Object.keys(header)[0]];
```
Comment (escardin) [inline angular2-jwt.ts]: Is this taking only the first entry of header? You know that you can have multiple headers of the same type right? Technically you should always be adding headers, regardless of if they're already set or not.

**Comment by chenkie** [comment]
Does this still require some adjusting?

File: angular2-jwt.ts (hunk 2 above)
Comment (CSchulz) [inline angular2-jwt.ts]: The Headers instance of providedOptions is used.

RequestOptions.merge replaces the headers completely and don't merge them.

Comment (CSchulz) [inline angular2-jwt.ts]: https://github.com/angular/angular/blob/master/packages/http/src/base_request_options.ts#L101

File: angular2-jwt.ts (hunk 3 above)
Comment (CSchulz) [inline angular2-jwt.ts]: I will change it.

File: angular2-jwt.ts (hunk 1 above)
Comment (CSchulz) [inline angular2-jwt.ts]: I will change it.

**Comment by CSchulz** [comment]
Tests added, review comments partly fixed.

please check.

**Comment by lsholubka** [comment]
Can this be merged please?

**Comment by chenkie** [comment]
Are you seeing any errors with Jasmine? I'm getting this even with the latest `@types/jasmine`

```
node_modules/@types/jasmine/index.d.ts(40,52): error TS1005: '=' expected.
node_modules/@types/jasmine/index.d.ts(42,62): error TS1005: '=' expected.
```

**Comment by CSchulz** [comment]
When do you get these errors? I have tried to reproduce them. I have even installed the dependencies again.

**Comment by CSchulz** [comment]
@chenkie Can you please give some more details about the errors?

**Pull request closed without merging**